# ⚡🏘 Household Power Consumption

Regressors used : Bagging Regressor, Voting Regressor, ExtraTree Regressor, Random Forest Regressor, Decision Tree Regressor

## Importing Essentials

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings

warnings.filterwarnings("ignore")

%matplotlib inline

Importing cleaned data from MongoDB

In [2]:
import pymongo

client = pymongo.MongoClient("mongodb+srv://Shaah_i:1Caps1low@shaahi.ku1bjvf.mongodb.net/?retryWrites=true&w=majority")
db = client.test
print(db)

Database(MongoClient(host=['ac-seysk7d-shard-00-02.ku1bjvf.mongodb.net:27017', 'ac-seysk7d-shard-00-00.ku1bjvf.mongodb.net:27017', 'ac-seysk7d-shard-00-01.ku1bjvf.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-jtsuo7-shard-0', tls=True), 'test')


In [3]:
database = client['machine_learning']
collection = database['household_power_consumption']

In [4]:
df = pd.DataFrame(list(collection.find()))

In [5]:
df

_id        Date      Time  Global_active_power  \
0      636b930628743111e118f442   8/10/2008  19:49:00                1.926   
1      636b930628743111e118f445   10/6/2010  08:29:00                1.802   
2      636b930628743111e118f456   21/1/2007  09:22:00                1.446   
3      636b930628743111e118f459   26/5/2010  06:14:00                0.342   
4      636b930628743111e118f461   15/1/2008  23:05:00                1.600   
...                         ...         ...       ...                  ...   
74084  636b930728743111e11a1574   16/6/2008  11:42:00                0.254   
74085  636b930728743111e11a1577   23/6/2009  17:51:00                0.330   
74086  636b930728743111e11a1582    6/7/2010  22:15:00                0.248   
74087  636b930728743111e11a1583   12/4/2010  21:22:00                1.390   
74088  636b930728743111e11a1584  17/10/2010  21:44:00                1.336   

       Global_reactive_power  Voltage  Global_intensity  Total_metering  Day  \
0                      0.074   240.95               8.0               1    8   
1                      0.082   238.38               7.6              18   10   
2                      0.086   237.92               6.0              18   21   
3                      0.228   241.27               1.8               0   26   
4                      0.000   243.69               6.6              19   15   
...                      ...      ...               ...             ...  ...   
74084                  0.144   238.45               1.2               0   16   
74085                  0.222   240.26               1.6               3   23   
74086                  0.058   241.19               1.0               1    6   
74087                  0.176   239.59               5.8               1   12   
74088                  0.128   242.63               5.4               0   17   

       Month  Year  
0         10  2008  
1          6  2010  
2          1  2007  
3          5  2010  
4          1  2008  
...      ...   ...  
74084      6  2008  
74085      6  2009  
74086      7  2010  
74087      4  2010  
74088     10  2010  

[74089 rows x 11 columns]

In [6]:
X = df.drop(["_id","Total_metering", "Date", "Time"],axis=1)
X

Global_active_power  Global_reactive_power  Voltage  Global_intensity  \
0                    1.926                  0.074   240.95               8.0   
1                    1.802                  0.082   238.38               7.6   
2                    1.446                  0.086   237.92               6.0   
3                    0.342                  0.228   241.27               1.8   
4                    1.600                  0.000   243.69               6.6   
...                    ...                    ...      ...               ...   
74084                0.254                  0.144   238.45               1.2   
74085                0.330                  0.222   240.26               1.6   
74086                0.248                  0.058   241.19               1.0   
74087                1.390                  0.176   239.59               5.8   
74088                1.336                  0.128   242.63               5.4   

       Day  Month  Year  
0        8     10  2008  
1       10      6  2010  
2       21      1  2007  
3       26      5  2010  
4       15      1  2008  
...    ...    ...   ...  
74084   16      6  2008  
74085   23      6  2009  
74086    6      7  2010  
74087   12      4  2010  
74088   17     10  2010  

[74089 rows x 7 columns]

In [7]:
y = df["Total_metering"]
y

0         1
1        18
2        18
3         0
4        19
         ..
74084     0
74085     3
74086     1
74087     1
74088     0
Name: Total_metering, Length: 74089, dtype: int64

**Train-Test split**

In [8]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size=0.3, random_state=32)

**Standardize the dataset**

In [9]:
# from sklearn.preprocessing import StandardScaler
# scaler = StandardScaler()

# X_train = scaler.fit_transform(X_train)
# X_test = scaler.transform(X_test)

**Creating model pipeline**

In [10]:
from sklearn.ensemble import RandomForestRegressor, BaggingRegressor, VotingRegressor, ExtraTreesRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression

In [11]:
Lin_reg = LinearRegression()
DT_reg = DecisionTreeRegressor()
RF_reg = RandomForestRegressor()
Bag_reg = BaggingRegressor(base_estimator=RF_reg, n_estimators=20, random_state=32)
ET_reg = ExtraTreesRegressor(n_estimators=20, random_state=32)
Vot_reg = VotingRegressor([('lr', Lin_reg), ('rf', RF_reg), ('r3', DT_reg)])

regressor = [Lin_reg, DT_reg, RF_reg, Bag_reg, ET_reg, Vot_reg]

# creating pipline
from sklearn.pipeline import Pipeline

for r in regressor:
    pipe = Pipeline(steps= [("regressor", r)])
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_test)
    
    mae = mean_absolute_error(y_test, y_pred)
    rsme = np.sqrt(mean_squared_error(y_test, y_pred))
    r_sq = r2_score(y_test, y_pred)
    adjr2 = 1-(1-r_sq)*(len(y_test)-1)/(len(y_test)-X_test.shape[1]-1)
    
    print(f"Algorithm: {r}")
    print(f"MAE: {mae}")
    print(f"RSME: {rsme}")
    print(f"R2: {r_sq}")
    print(f"Adjusted R2: {adjr2}")    
    print("\n")

Algorithm: LinearRegression()
MAE: 4.210219712283747
RSME: 6.66061331325282
R2: 0.7220002693331868
Adjusted R2: 0.7219126867185477


Algorithm: DecisionTreeRegressor()
MAE: 3.09915868088361
RSME: 7.435796186088076
R2: 0.6535258104244621
Adjusted R2: 0.6534166552272422


Algorithm: RandomForestRegressor()
MAE: 2.6227736300981004
RSME: 5.409018237840687
R2: 0.8166618585380014
Adjusted R2: 0.8166040986482569


Algorithm: BaggingRegressor(base_estimator=RandomForestRegressor(), n_estimators=20,
                 random_state=0)
MAE: 2.665338379427849
RSME: 5.3949593491055134
R2: 0.8176136692788825
Adjusted R2: 0.8175562092530015


Algorithm: ExtraTreesRegressor(n_estimators=20, random_state=0)
MAE: 2.5887580420209653
RSME: 5.497450540927882
R2: 0.810618045389768
Adjusted R2: 0.8105583814227906


Algorithm: VotingRegressor(estimators=[('lr', LinearRegression()),
                            ('rf', RandomForestRegressor()),
                            ('r3', DecisionTreeRegressor())])
MAE: 3.0

Here we can see that by using ensemble techniques the r2 score has increased by around 16% from lowest one.